### ISSUE TITLE
Different results with -j4 and -j8

### ISSUE THREAD

**Issue by nedbat** [pylint-dev/pylint#4356]
I'm not sure if this is a duplicate of #2573 or #374 because I don't know if those are particular to the specific checks they mention.

### Steps to reproduce

1. `git clone https://github.com/nedbat/coveragepy; cd coveragepy`
2. `git checkout 50cc68846f`
3. `pip install tox`
4. `tox -e lint --notest`
5. `.tox/lint/bin/python -m pylint --notes= -j 4 coverage tests doc ci igor.py setup.py __main__.py`
6. No violations are reported
7. `.tox/lint/bin/python -m pylint --notes= -j 8 coverage tests doc ci igor.py setup.py __main__.py`
```
************* Module coverage.pytracer
coverage/pytracer.py:140 E: self.should_start_context is not callable (not-callable)
coverage/pytracer.py:165 E: self.should_trace is not callable (not-callable)
coverage/pytracer.py:166 E: 'self.should_trace_cache' does not support item assignment (unsupported-assignment-operation)
coverage/pytracer.py:171 E: Value 'self.data' doesn't support membership test (unsupported-membership-test)
coverage/pytracer.py:172 E: 'self.data' does not support item assignment (unsupported-assignment-operation)
coverage/pytracer.py:173 E: Value 'self.data' is unsubscriptable (unsubscriptable-object)
coverage/pytracer.py:259 E: self.warn is not callable (not-callable)
```

A different manifestation of what might be the same underlying cause: in my CI, running -j4, I get occasional failures with these same violations reported.

### Expected behavior

The number of processes shouldn't change the results, only the performance.

### pylint --version output

Result of `pylint --version` output:

```
$ .tox/lint/bin/python -m pylint --version
pylint 2.7.4
astroid 2.5.3
Python 3.8.8 (default, Mar 21 2021, 13:50:20)
[Clang 12.0.0 (clang-1200.0.32.29)]
```

**Comment by Pierre-Sassoulas** [comment]
Closing as duplicate of #374

**Comment by jacobtylerwalls** [comment]
Thanks for the beautiful reproducer. Working on a fix for this that seems distinct from the freeform discussion in #374.

**Comment by jacobtylerwalls** [comment]
Regression in 55076ca06018dd3a5d2748610c5aa81efbb96563.

**Comment by jacobtylerwalls** [comment]
Never mind, the indeterminacy goes a little deeper than that commit. PR to come, with luck.
